### PR TITLE
Lock Poisoning

### LINKED ISSUE #10: vicky: Implement Lock Poisoning
We want to poison a lock, if a task did not succeed.

If a lock is poisoned, it cannot be used and every other task, which wants to claim the lock, must be stuck.
A lock can only be resolved manually by an engineer.

### PR DESCRIPTION
**Pull request by Kek5chen**
Closes #10

### CODE REVIEW

**Comment by Kek5chen** [comment]
@johannwagner should a poisoned read lock also not allow another read lock to be used for the same resource? I don't see an issue with this.

File: vicky/src/bin/vicky/tasks.rs [hunk 1]
```diff
@@ -217,6 +221,12 @@ pub async fn tasks_finish(
     log_drain.finish_logs(&id).await?;
 
     task.status = TaskStatus::FINISHED(finish.result.clone());
+    
+    if finish.result == TaskResult::ERROR {
+        task.locks.iter_mut().for_each(|lock| lock.poison(&task.id));
+    }
+    
+    // TODO: this clone is weird and can be saved I think
```
Comment (johannwagner) [inline vicky/src/bin/vicky/tasks.rs:229]: Then try to remove it? :thinking:

File: vicky/src/lib/vicky/scheduler.rs [hunk 2]
```diff
@@ -343,8 +355,65 @@ mod tests {
...
+        let res = Scheduler::new(tasks, &poisoned_locks, &[]).unwrap();
+
+        assert_eq!(
+            res.get_next_task().unwrap().display_name,
+            "I need to test something"
+        );
+    }
+
+    #[test]
+    fn schedule_different_tasks_with_poisoned_lock_ro() {
```
Comment (johannwagner) [inline vicky/src/lib/vicky/scheduler.rs:405]: This is the same test as `schedule_with_poisoned_lock`

File: vickyctl/src/humanize.rs [hunk 3]
```diff
@@ -1,9 +1,23 @@
...
 use std::io::{ErrorKind, Write};
 use std::process::{Command, Stdio};
 
+use log::debug;
+use which::which;
+
+use crate::AppContext;
+use crate::error::Error;
+
+pub fn ensure_jless(needed_for: &str) -> Result<(), Error> {
```
Comment (johannwagner) [inline vickyctl/src/humanize.rs:10]: Not sure, why this change is in this PR, but it can stay.

File: vickyctl/src/main.rs [hunk 4]
```diff
@@ -61,11 +62,23 @@ struct TasksArgs {
     ctx: AppContext,
 }
 
+#[derive(Args, Debug)]
+#[command(version, about = "Show all poisoned locks vicky is managing", long_about = None)]
+struct LocksArgs {
+    #[command(flatten)]
+    ctx: AppContext,
+    #[clap(long)]
```
Comment (johannwagner) [inline vickyctl/src/main.rs:70]: active and poisoned should not be a option but an argument.

Therefore valid commands would be `... locks active` and not `... locks --active`.
This eliminates the needed validation that at least `active` or `poisoned` is given as argument, which is currently missing.

**Review by johannwagner — CHANGES_REQUESTED** [review]

File: vicky/src/bin/vicky/tasks.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kek5chen) [reply vicky/src/bin/vicky/tasks.rs:229]: It's not that easy. But it's noted.

File: vicky/src/lib/database/entities/lock.rs [hunk 5]
```diff
@@ -1,37 +1,59 @@
...
 pub enum Lock {
-    WRITE { name: String },
-    READ { name: String },
+    WRITE {
+        name: String,
+        poisoned: Option<Uuid>,
+    },
+    READ {
+        name: String,
+        poisoned: Option<Uuid>,
```
[the remark this answers is not in the document]
Comment (Kek5chen) [reply vicky/src/lib/database/entities/lock.rs:13]: That's true, but the wrapper functions are the way to access these anyways, and in the code, it really only needs to check if it is poisoned or not. Also since this is an enum, the math statements get exponentially longer the longer the class member names. You are right, but I'd still rather keep it like this.

File: vicky/src/lib/vicky/scheduler.rs (hunk 2 above)
Comment (Kek5chen) [reply vicky/src/lib/vicky/scheduler.rs:405]: nope

File: vickyctl/src/humanize.rs (hunk 3 above)
Comment (Kek5chen) [reply vickyctl/src/humanize.rs:10]: Well I added some stuff to the cli and refactored this out since it's now used more than once, while before it wasn't.

File: vickyctl/src/main.rs (hunk 4 above)
Comment (Kek5chen) [reply vickyctl/src/main.rs:70]: Well, we don't really care for active locks all that much. So poisoned locks is the default, which is why this is the case. If no argument is supplied, it shows poisoned locks. And to me, active or poisoned are properties of locks, therefore I made them arguments.

File: vicky/src/lib/vicky/scheduler.rs (hunk 2 above)
Comment (Kek5chen) [reply vicky/src/lib/vicky/scheduler.rs:405]: a read lock can be scheduled with N other read locks, but not when one other poisoned read lock exists, while write locks shouldn't be scheduled together either way.

File: vicky/src/lib/database/entities/lock.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (johannwagner) [reply vicky/src/lib/database/entities/lock.rs:13]: I would be fine with `poisonedBy`, otherwise it feels like a boolean.

**Comment by johannwagner** [comment]
The fix for the tests is not cherry-pickable easily.
Therefore, we could close #41 and merge it with #42?
